lib/internal/fs/streams.js

```diff
@@ -74,6 +74,7 @@ function ReadStream(path, options) {
   this.fd = options.fd === undefined ? null : options.fd;
   this.flags = options.flags === undefined ? 'r' : options.flags;
   this.mode = options.mode === undefined ? 0o666 : options.mode;
+  this.fs = options.fs || fs;
```

Comment:
If we did go with this approach, what I would prefer is that we grab the `open`, `read`, and `close` function references off the `fs` object rather than the `fs` object itself. e.g...

```js
const kOpen = Symbol('kOpen');
const kClose = Symbol('kClose');
const kRead = Symbol('kRead');

/* then in the constructor... */
this[kRead] = (options.fs || fs).read;
this[kClose] = (options.fs || fs).close;
this[kRead] = (options.fs || fs).read;
```

This prevents later monkeypatching from altering the behavior of the already created object.

Reply:
There should also be validation that the provided `read`, `close`, and `open` are actually functions.